common/softassert/softassert.go

```diff
@@ -0,0 +1,50 @@
+// The MIT License
+//
+// Copyright (c) 2025 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package softassert
+
+import (
+	"fmt"
+
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+)
+
+// That performs a soft assertion by logging an error if the given condition is false.
+// It is meant to indicate a condition that are always expected to be true.
+// Returns true if the condition is met, otherwise false.
+//
+// Example:
+// softassert.That(logger, object.state == "ready", "object is not ready: %v", object.state)
+//
+// Best practices:
+// - Use it to check for programming errors and invariants.
+// - Use it to communicate assumptions about the code.
+// - Use it to abort or recover from an unexpected state.
+// - Never use it as a substitute for regular error handling, validation, or control flow.
+func That(logger log.Logger, condition bool, format string, args ...any) bool {
+	if !condition {
+		// By using the same prefix for all assertions, they can be reliably found in logs.
+		logger.Error("failed assertion: "+fmt.Sprintf(format, args...), tag.FailedAssertion())
```

Comment:
similarly, in the interest of keeping the code as small as possible, I'd get rid of the string prefix

Reply:
Alright. I think we lose a bit of discovery there but the testlogger prefix and the logger tag hopefully make up for that.

Reply:
Hm, on second thought, I like how explicit and clear it is with the prefix. Since it's only prefixing on a failed assertion, that shouldn't be too often.

Reply:
it's more about the code size (which determines whether this call can be inlined) than the run time when the condition is false (which should be never of course)

Reply:
on that note, can we make `tag.FailedAssertion` a static value instead of a function, so it's just a load instead of a function call?